internal/grpc/commander/connection.go

```diff
@@ -0,0 +1,203 @@
+package commander
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/go-logr/logr"
+	"github.com/nginx/agent/sdk/v2/grpc"
+	"github.com/nginx/agent/sdk/v2/proto"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/grpc/commander/exchanger"
+)
+
+// State is the state of the connection.
+type State int
+
+const (
+	// StateConnected means the connection is active (connected) but is not registered.
+	StateConnected State = iota
+	// StateRegistered means the connection is active and registered.
+	StateRegistered
+	// StateInvalid means the connection is active and attempted to register, but its registration info was invalid.
+	// We cannot push config to a connection in this state.
+	StateInvalid
+)
+
+// connection represents a connection to an agent.
+type connection struct {
+	cmdExchanger exchanger.CommandExchanger
+	logger       logr.Logger
+	id           string
+	nginxID      string
+	systemID     string
+	state        State
+}
+
+func (c *connection) ID() string {
+	return c.id
+}
+
+func (c *connection) State() State {
+	return c.state
+}
+
+// ReceiveFromUploadServer uploads data chunks from the UploadServer and logs them.
+// FIXME(kate-osborn): NKG doesn't need this functionality and ideally we wouldn't have to implement and maintain this.
+// Figure out how to remove this without causing errors in the agent.
+func (c *connection) ReceiveFromUploadServer(server proto.Commander_UploadServer) error {
```

Comment:
> does this loop terminate when all chucks are received?

yes. I'll add a comment

> do we need to propagate context here to terminate the loop earlier if we're shutting down?

I don't think so. If we shut down, the grpc server is stopped, which closes all connections. `Recv()` will terminate with an error, and we will exit.

Reply:
> If we shut down, the grpc server is stopped, which closes all connections. Recv() will terminate with an error, and we will exit.

is this the case why it makes sense to gracefully shutdown the server (`GracefulStop()`) ?

Reply:
I saw issues with `GracefulStop()` where the manager timed out waiting for connections to close. I think it's because the agent always tries to reconnect after any server error. 

```
// Stop stops the gRPC server. It immediately closes all open
// connections and listeners.
// It cancels all active RPCs on the server side and the corresponding
// pending RPCs on the client side will get notified by connection
// errors.
```

Reply:
Sure, I can look into why `GracefulStop()` causes issues after I finish the config piece.

Reply:
I'll add a FIXME so I don't  forget